### ISSUE TITLE
Unable to use weaver.ignore in Simplified

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#3926]
I am trying to use a nuget package inside the Simplified shared library. The package has to be excluded from weaver.

This does not work.

Steps to reproduce:

- Open [Simplified](https://github.com/StarcounterPrefabs/Simplified) project in Visual Studio.
- Install [Scrypt.NET](https://github.com/viniciuschiele/scrypt) nuget package.
- Rebuild the solution.
- Open [Launcher](https://github.com/StarcounterPrefabs/Launcher) in Visual Studio.
- Try to run Launcher. 
 - **Fails.** See the snippet below.
- Add `weaver.ignore` file to both solutions with `Scrypt.dll` entry.
- Try to run Launcher once again.
 - **Fails again**. See the snipped below.

```
The weaver failed to resolve a reference to an assembly. Referenced assembly: Scrypt, Version=1.1.0.0, Culture=neutral. Probable referrer: Simplified.Data.Model.dll (in C:\Program Files\Starcounter\EditionLibraries). (ScErrWeaverFailedResolveReference (SCERR2147))
```

**Note:** if an app (not Simplified) contains the `Scrypt.NET` reference, then adding `weaver.ignore` solves the problem.

cc @per-samuelsson

**Comment by per-samuelsson** [comment]
I've discussed this with @miyconst off-tracker, and it seems the use-case is not supported. I think simply because it was never tried - there have never been a need to exclude anything from one of our "special" assemblies, like "Simplified".

I need to consider what options we have here. In the perfect world, I would most of all like to nuke these "special assemblies" that was introduced some years ago, because they add maintenance and weird problems like this. If we can't do that (in reasonable time), I guess we need to come up with some kind of additional exclusion policy that embrace not only user assemblies, but our "special assemblies" too.

**Comment by alemoi** [comment]
Yeah, when it comes to Simplified - the original idea was that its an exception, meaning that there will be no other libraries, or they have to be packaged the same way as Simplified.

**Comment by per-samuelsson** [comment]
> Yeah, when it comes to Simplified - the original idea was that its an exception, meaning that there will be no other libraries, or they have to be packaged the same way as Simplified.

I've now set this up locally, and it turns out this is specific to "special assembly" Simplified. That project is setup in quite some unconvential way:

1. When it builds, it builds -> %StarcounterBin%\Public Assemblies (where `Bin\[Platform]\[Config]` would be the expected.
2. After it has built, in a post-build step, it creates folder %StarcounterBin%\EditionLibraries and copies `$(AssemblyName).dll+pdb` into that directory.

When the weaver is told to include `Simplified.dll` in `%StarcounterBin%\EditionLibraries` as a reference to every application ever built, it will of course fail resolving a reference to any library that is not to be found anywhere (in this case, `Scrypt.dll`).

I proposed, as a workaround, just to help getting passed this, to copy `Scrypt.dll` in the same maner. That works. But all and all, I think it would be really good to take a look at how these "special assemblies" are set up and try not so many "hacks".

**Comment by alemoi** [comment]
Thanks for overview, @per-samuelsson Probably it then should be just in "EditionLibraries" not in "Public Assemblies". Should I fix it or you are on it?

**Comment by per-samuelsson** [comment]
> Should I fix it or you are on it?

I didn't intend doing anything more to it, no.

> Probably it then should be just in "EditionLibraries" not in "Public Assemblies"

I can't tell, really, since I haven't been really involved in designing this with "special built in assemblies" (we have `LibrariesWithDatabaseClasses` that is "special" too). I don't really have an opinion, other than I think it's problematic in that sense that it's unconventional as a whole and that no one seem to own it really.

As for those two folders, "Public Assemblies" is a Visual Studio pattern and the enabler for VS to find assemblies recommended by us to reference, so it comes with semantics kind of. "EditionLibraries" is just something kind of "under the hood".

Bottomline is that if we are to support these things, we must make sure references being introduced don't mess up, so wherever you want our runtime (weaver and host) to look for special, built-in assemblies, it's important that those assemblies referenced by those special ones turn up there too. Or else there will be trouble.

**Comment by per-samuelsson** [comment]
(By the way, IMO we should move this issue into Simplified issue tracker)

**Comment by alemoi** [comment]
@per-samuelsson thanks for the info!

@miyconst then we need to decide together with @dan31  if your library should be a part of Starcounter distribution.

**Comment by miyconst** [comment]
@per-samuelsson, @alemoi thank you for the investigations. I will just copy the assembly to `EditionLibraries` for now.

**Comment by alemoi** [comment]
> I will just copy the assembly to EditionLibraries for now.

For your local environment sure. But let's discuss with @dan31 first, before putting it in SC distribution.

By the way, let us know if just putting in `EditionLibraries` works.

**Comment by miyconst** [comment]
>For your local environment sure. But let's discuss with @dan31 first, before putting it in SC distribution.

Sure, I need to keep development for now :).

>By the way, let us know if just putting in `EditionLibraries` works.

It does.

**Comment by per-samuelsson** [comment]
Just out of curiosity: what would be bad in putting `Simplified` artifacts in a NuGet package and have developers in need of it just bring it down that way? 

With that, users would be in more control, we would be more transparent AND we would not need to maintain any weird special treatment for at least that library any more.

**Comment by alemoi** [comment]
So you mean, @per-samuelsson, that each app will come with a specific version of Simplified? I wonder if that would not cause problems if different versions are used simultaneously by different apps deployed to same database. What do you think?

**Comment by per-samuelsson** [comment]
> So you mean, @per-samuelsson, that each app will come with a specific version of Simplified? I wonder if that would not cause problems if different versions are used simultaneously by different apps deployed to same database. What do you think?

It's really not my area of expertise. But if such a thing should go wrong, I think we need to treat that as a high-prio bug. Rationale: how could we ever be in control over what Starcounter libraries are used by different applications and their different versions? In my world, app isolation and virtual schemas was there to support exactly that, and if they can't (long term), we'll be in trouble.

IMO we **MUST** start thinking as every single application as living in total isolation, and design our bits after that. Anything that don't work because of a technical implementation detail (such as hosting several apps in one process) has to be considered a bug that need to be fixed.

**Comment by miyconst** [comment]
@per-samuelsson, @alemoi,

Currently we do not support multiple versions of a single assembly loaded into one code host. This is a known bug/limitation. This is the main reason why Simplified is distributed with Starcounter instead of a nuget package.

Making it be a nuget package creates much more problems than distributing it with Starcounter and dealing with weaver.

>IMO we MUST start thinking as every single application as living in total isolation, and design our bits after that. Anything that don't work because of a technical implementation detail (such as hosting several apps in one process) has to be considered a bug that need to be fixed.

I think we already do so. In Starcounter 3.0 every app will have it's own isolated host, and it should be possible to load different version of same assembly for different apps.

cc @eriohl, @bigwad, @linkdata.

**Comment by linkdata** [comment]
> In Starcounter 3.0 every app will have it's own isolated host, and it should be possible to load different version of same assembly for different apps.

That's one of the hosting models, and in that model that will be possible.
